<a href="https://colab.research.google.com/github/marianagg0412/Analisis/blob/main/Ra%C3%ADces_de_ecuaciones_no_lineales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Raíces de ecuaciones no-lineales**



## **Método de Bisección**

In [3]:
import numpy as np
import matplotlib.pyplot as plt

In [22]:
#épsilon de la máquina
print(np.finfo(float))

Machine parameters for float64
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
smallest_normal = 2.2250738585072014e-308   smallest_subnormal = 4.9406564584124654e-324
---------------------------------------------------------------



Machine parameters for float64
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
smallest_normal = 2.2250738585072014e-308   smallest_subnormal = 4.9406564584124654e-324

---------------------------------------------------------------

In [23]:
eps = np.finfo(float).eps
print(eps)

2.220446049250313e-16


In [4]:
def func(x):
  y = 3*x**2-np.exp(x)
  #y = 9700*(1-np.exp(-x/15))-480*x
  #y = (420/(x/12))*((1+x/12)**250 - 1)-300000
  return y

In [ ]:
print(func(-1))
print(func(0))

2.6321205588285577
-1.0


In [5]:
def biseccion(a,b, tolerancia, delta, Nmax): #intervalo cambio de signo, criterios de parada (tolerancia y delta)
  i = 0 # cont, iteraciones
  d = delta + 1 #while delta >, para entrar la primera vez al ciclo, abs(f(y))
  error = tolerancia + 1 # Error < épsilon
  s = np.array([]) #solución
  aux = a #aux variable for error slot in first iteration
  while (error > tolerancia) and (delta < d) and (Nmax > i):
    i+=1
    x = (a+b)/2 #punto medio
    y = func(x) #function eval-ed in x
    d = abs(y) #abs y
    error = abs(x-aux) #Absolute value
    #error = abs(x-aux)/abs(x) #Relative
    row = np.array([i,x,y,error])
    s = np.append(s,row)

    if(y*func(a) < 0):
      b = x
    else:
      a = x

    aux = x #So that error is calculated with respect to the x from right before
  print(s)

In [6]:
biseccion(-1,0,0,0,200) #Para por error
#biseccion(0,1,0,0,200) #Para por delta
#biseccion(3,4,0,0,200) #Para por error

[ 1.00000000e+00 -5.00000000e-01  1.43469340e-01  5.00000000e-01
  2.00000000e+00 -2.50000000e-01 -5.91300783e-01  2.50000000e-01
  3.00000000e+00 -3.75000000e-01 -2.65414279e-01  1.25000000e-01
  4.00000000e+00 -4.37500000e-01 -7.14297764e-02  6.25000000e-02
  5.00000000e+00 -4.68750000e-01  3.33956779e-02  3.12500000e-02
  6.00000000e+00 -4.53125000e-01 -1.96718770e-02  1.56250000e-02
  7.00000000e+00 -4.60937500e-01  6.69804230e-03  7.81250000e-03
  8.00000000e+00 -4.57031250e-01 -6.52786306e-03  3.90625000e-03
  9.00000000e+00 -4.58984375e-01  7.48508301e-05  1.95312500e-03
  1.00000000e+01 -4.58007812e-01 -3.22906552e-03  9.76562500e-04
  1.10000000e+01 -4.58496094e-01 -1.57774723e-03  4.88281250e-04
  1.20000000e+01 -4.58740234e-01 -7.51608177e-04  2.44140625e-04
  1.30000000e+01 -4.58862305e-01 -3.38418668e-04  1.22070312e-04
  1.40000000e+01 -4.58923340e-01 -1.31793918e-04  6.10351562e-05
  1.50000000e+01 -4.58953857e-01 -2.84740435e-05  3.05175781e-05
  1.60000000e+01 -4.58969

## **Método de regla falsa**

In [ ]:
func(0.08)

-31286.910824706254

In [7]:
def regla_falsa(a,b, tolerancia, delta, Nmax): #En general se demora menos que bisección
  i = 0 # cont, iteraciones
  d = delta + 1 #while delta >, para entrar la primera vez al ciclo, abs(f(y))
  error = tolerancia + 1 # Error < épsilon
  s = np.array([]) #solución
  aux = a #aux variable for error slot in first iteration
  while (error > tolerancia) and (delta < d) and (Nmax > i):
    i+=1
    x = a - ((func(a)*(b-a))/(func(b) - func(a)))
    y = func(x) #function eval-ed in x
    d = abs(y) #abs y
    error = abs(x-aux) #Absolute error
    #error = abs(x-aux)/abs(x) #relative error
    row = np.array([i,x,y,error])
    s = np.append(s,row)

    if(y*func(a) < 0):
      b = x
    else:
      a = x

    aux = x #So that error is calculated with respect to the x from right before
  print(s)

In [8]:
regla_falsa(-1,0,0,0,200) #Para por error
#regla_falsa(0,1,0,0,200) #Para por delta
#regla_falsa(3,4,0,0,200) #Para por error

[ 1.00000000e+00 -2.75321258e-01 -5.31922759e-01  7.24678742e-01
  2.00000000e+00 -3.97150537e-01 -1.99047176e-01  1.21829279e-01
  3.00000000e+00 -4.39534279e-01 -6.47652856e-02  4.23837421e-02
  4.00000000e+00 -4.52993774e-01 -2.01120132e-02  1.34594954e-02
  5.00000000e+00 -4.57141750e-01 -6.15485455e-03  4.14797577e-03
  6.00000000e+00 -4.58408188e-01 -1.87512864e-03  1.26643851e-03
  7.00000000e+00 -4.58793745e-01 -5.70492664e-04  3.85556561e-04
  8.00000000e+00 -4.58911022e-01 -1.73495498e-04  1.17277034e-04
  9.00000000e+00 -4.58946685e-01 -5.27559383e-05  3.56633794e-05
  1.00000000e+01 -4.58957530e-01 -1.60412338e-05  1.08441853e-05
  1.10000000e+01 -4.58960827e-01 -4.87752044e-06  3.29731699e-06
  1.20000000e+01 -4.58961830e-01 -1.48306055e-06  1.00258505e-06
  1.30000000e+01 -4.58962134e-01 -4.50939413e-07  3.04846186e-07
  1.40000000e+01 -4.58962227e-01 -1.37112600e-07  9.26915201e-08
  1.50000000e+01 -4.58962255e-01 -4.16904413e-08  2.81837743e-08
  1.60000000e+01 -4.58962

No importa el método que utilicemos, generalmente darán lo mismo cuando los criterios de parada son nulos.
Generalmente bisección requerirá más iteraciones que regla falsa.

**Ejercicio: ** La siguiente función representa una aproximación de la tasa de interés anual I con la que se conseguiría un capital acumulado A tras hacer 250 depósitos mensuales:

$$A = (\dfrac{12P}{I}){(1 + \dfrac{I}{12})^{250}-{1}}$$

Use el método de regla falsa para determinar la tasa de interés, con una calidad de 0.001%, si la cantidad depositada es P = $420 USD y el capital esperado es de $300.000 USD

300000 = (420/(I/12))((1 + (I/12)^250 -1)
0 = (420/(I/12))((1 + (I/12)^250 -1) - 300000 = (F(I))

Tasa de interés (0;1]
Criterios de parada:
I: interés -> Ea, Er, E% (Tolerancia)
f(I): capital -> Dólares (Delta)


In [ ]:
#regla_falsa(0.08, 0.09, 0.00001,0,200)
biseccion(0.08,0.09,10**-5,0,200)

[ 1.00000000e+00  8.50000000e-02 -1.30648439e+04  5.00000000e-03
  2.00000000e+00  8.75000000e-02 -3.41289611e+03  2.50000000e-03
  3.00000000e+00  8.87500000e-02  1.55493760e+03  1.25000000e-03
  4.00000000e+00  8.81250000e-02 -9.41005195e+02  6.25000000e-04
  5.00000000e+00  8.84375000e-02  3.03944170e+02  3.12500000e-04
  6.00000000e+00  8.82812500e-02 -3.19284070e+02  1.56250000e-04
  7.00000000e+00  8.83593750e-02 -7.85858320e+00  7.81250000e-05
  8.00000000e+00  8.83984375e-02  1.47995605e+02  3.90625000e-05
  9.00000000e+00  8.83789062e-02  7.00567174e+01  1.95312500e-05
  1.00000000e+01  8.83691406e-02  3.10961192e+01  9.76562500e-06]


## **Método de Punto Fijo**

Punto fijo no inicia directamente en el intervalo [a,b]

In [9]:
def func_g(x):
  y = -np.sqrt(np.exp(x)/3)
  #y = 970/48*((1-np.exp(-x/15)))
  return y

In [ ]:
func_g(8)

8.353190980734773

In [10]:
def punto_fijo(x0, tolerancia, delta, Nmax): #intervalo cambio de signo, criterios de parada (tolerancia y delta)
  i = 0 # cont, iteraciones
  d = delta + 1 #while delta >, para entrar la primera vez al ciclo, abs(f(y))
  error = tolerancia + 1 # Error < épsilon
  s = np.array([]) #solución
  while (error > tolerancia) and (delta < d) and (Nmax > i):
    i+=1
    x = func_g(x0)
    y = func(x) #function eval-ed in x
    d = abs(y) #abs y, delta
    error = abs(x-x0) #Absolute error
    #error = abs(x-x0)/abs(x) #Relative error
    row = np.array([i,x,y,error])
    s = np.append(s,row)

    x0 = x #renewing x
  print(s)

In [31]:
punto_fijo(-0.5,0,0,200)
#Comenzamos en el punto medio, to compare, since there's a theoretical homogenous distribution
#Haremos los otros intervalos

[ 1.00000000e+00 -4.49640842e-01 -3.13265424e-02  5.03591582e-02
  2.00000000e+00 -4.61106351e-01  7.27159205e-03  1.14655097e-02
  3.00000000e+00 -4.58470504e-01 -1.66431975e-03  2.63584713e-03
  4.00000000e+00 -4.59075132e-01  3.82160100e-04  6.04627421e-04
  5.00000000e+00 -4.58936368e-01 -8.76864125e-05  1.38763730e-04
  6.00000000e+00 -4.58968211e-01  2.01230156e-05  3.18429658e-05
  7.00000000e+00 -4.58960904e-01 -4.61781883e-06  7.30739636e-06
  8.00000000e+00 -4.58962581e-01  1.05970407e-06  1.67690768e-06
  9.00000000e+00 -4.58962196e-01 -2.43181998e-07  3.84818777e-07
  1.00000000e+01 -4.58962284e-01  5.58056856e-08  8.83086440e-08
  1.10000000e+01 -4.58962264e-01 -1.28063518e-08  2.02651680e-08
  1.20000000e+01 -4.58962268e-01  2.93881619e-09  4.65047373e-09
  1.30000000e+01 -4.58962267e-01 -6.74402645e-10  1.06719594e-09
  1.40000000e+01 -4.58962268e-01  1.54762647e-10  2.44901266e-10
  1.50000000e+01 -4.58962268e-01 -3.55151464e-11  5.62002112e-11
  1.60000000e+01 -4.58962

**Ejemplo 2**
Proyectil

0 = 9700(1-e^(-t/15)) - 480t -> f(t)

g(t) = (970*(1-e^(-t/15)))/48 -> función g

In [ ]:
punto_fijo(9,10**-3,10**-4,200)

[1.00000000e+00 9.11776485e+00 4.16308979e+01 1.17764854e-01
 2.00000000e+00 9.20449589e+00 3.04518534e+01 8.67310374e-02
 3.00000000e+00 9.26793725e+00 2.21634535e+01 6.34413612e-02
 4.00000000e+00 9.31411111e+00 1.60721560e+01 4.61738614e-02
 5.00000000e+00 9.34759477e+00 1.16240536e+01 3.34836583e-02
 6.00000000e+00 9.37181155e+00 8.39084554e+00 2.42167783e-02
 7.00000000e+00 9.38929248e+00 6.04853490e+00 1.74809282e-02
 8.00000000e+00 9.40189359e+00 4.35571204e+00 1.26011144e-02
 9.00000000e+00 9.41096799e+00 3.13439943e+00 9.07440009e-03
 1.00000000e+01 9.41749799e+00 2.25436181e+00 6.52999881e-03
 1.10000000e+01 9.42219458e+00 1.62080353e+00 4.69658710e-03
 1.20000000e+01 9.42557125e+00 1.16498479e+00 3.37667402e-03
 1.30000000e+01 9.42799830e+00 8.37194024e-01 2.42705165e-03
 1.40000000e+01 9.42974246e+00 6.01549817e-01 1.74415422e-03
 1.50000000e+01 9.43099569e+00 4.32188978e-01 1.25322879e-03
 1.60000000e+01 9.43189608e+00 3.10487841e-01 9.00393705e-04]


#**Método de Newton-Raphson**

f(x) = 0 -> Continua [a,b] , f(a)*f(b) < 0


In [13]:
func(0)

-1.0

In [15]:
def derivada(x):
  y = 6*x - np.exp(x)
  return y

In [16]:
derivada(0)

-1.0

In [49]:
#Funcion
def funcion(x):
  y = 300 - ((0.25*32.17)/0.1)*x + (((0.25**2)*32.17)/(0.1**2))*(1-np.exp((-0.1*x)/0.25))
  return y

#Derivada

def derivative(x):
  y = -(0.25*32.17/0.1) + (((0.25**2)*32.17)/(0.1**2))*((0.1/0.25)*np.exp(-0.1*x/0.25))
  return y

In [48]:
def newton_raphson(x0, tolerancia, delta, Nmax): #intervalo cambio de signo, criterios de parada (tolerancia y delta)
  i = 0 # cont, iteraciones
  d = delta + 1 #while delta >, para entrar la primera vez al ciclo, abs(f(y))
  error = tolerancia + 1 # Error < épsilon
  s = np.array([]) #solución
  while (error > tolerancia) and (delta < d) and (Nmax > i):
    i+=1
    x = x0 - (funcion(x0)/derivative(x0))
    y = funcion(x) #function eval-ed in x
    d = abs(y) #abs y, delta
    error = abs(x-x0) #Absolute error
    #error = abs(x-x0)/abs(x) #Relative error
    row = np.array([i,x,y,error])
    s = np.append(s,row)

    x0 = x #renewing x
  print(s)

In [29]:
newton_raphson(-0.5,eps,eps,200) #Falla por Nmax in this setting
#El error y la función evaluada en x (y) está estancado
#Debemos flexibilizar los métodos de parada, utilizaremos el épsilon de la máquina para delta y la tolerancia

#Cuando este método converge, lo hará muy rápido

[ 1.00000000e+00 -4.60219570e-01  4.26110821e-03  3.97804300e-02
  2.00000000e+00 -4.58963518e-01  4.23492102e-06  1.25605169e-03
  3.00000000e+00 -4.58962268e-01  4.19919655e-12  1.25081866e-06
  4.00000000e+00 -4.58962268e-01  1.11022302e-16  1.24028565e-12]


Cuando nos pidan los criterios más exigentes de parada, utilizaremos el épsilon de la máquina.

Cuando la recta es muy plana, los valores se comienzan a salir del intervalo.

**Ejercicio de física**

s(t) = s0 - mg/kt + m^2g/k^2(1-e^(-kt/m))

Función:

0 = 300 - (0.25*32.17)/0.1t + ((0.25^2*32.17)/0.1^2)(1-exp^(-0.1t/0.25) -> = f(t)

Intervalo: continua en todos los reales pero es tiempo por tanto
[0;inf)

Cambio de signo
[6;6.1]

t: tiempo (s) Tolerancia
f(t): altura (ft) Delta

Ea = 10^-3s
Delta = eps



In [33]:
funcion(6)

0.2725215167475312

In [45]:
derivative(6)

-73.129008606699

In [50]:
newton_raphson(6.05,10**(-3),eps,200) #6.05 midpoint between 6 and 6.1

[ 1.00000000e+00  6.00376837e+00 -3.07601277e-03  4.62316347e-02
  2.00000000e+00  6.00372631e+00 -2.57708166e-09  4.20565061e-05]
